From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NzSN <aydenlin@foxmail.com>
Date: Thu, 4 Jul 2024 19:52:04 +0800
Subject: Unblock access of mixed contents


diff --git a/third_party/blink/renderer/core/loader/mixed_content_checker.cc b/third_party/blink/renderer/core/loader/mixed_content_checker.cc
index 087986ad7ee53cf9a3c056e21066660c5baada6a..cfe3dcdcca729e466b131d4c06041af33416d88f 100644
--- a/third_party/blink/renderer/core/loader/mixed_content_checker.cc
+++ b/third_party/blink/renderer/core/loader/mixed_content_checker.cc
@@ -444,6 +444,10 @@ bool MixedContentChecker::ShouldBlockFetch(
     policy->ReportMixedContent(url_before_redirects, redirect_status);
 
   Settings* settings = mixed_frame->GetSettings();
+  if (settings && settings->GetAllowRunningOfInsecureContent()) {
+    return false;
+  }
+
   auto& local_frame_host = frame->GetLocalFrameHostRemote();
   WebContentSettingsClient* content_settings_client =
       frame->GetContentSettingsClient();
@@ -578,6 +582,7 @@ bool MixedContentChecker::ShouldBlockFetch(
       allowed ? MixedContentResolutionStatus::kMixedContentWarning
               : MixedContentResolutionStatus::kMixedContentBlocked,
       devtools_id);
+
   return !allowed;
 }
 
